controller/cmd/service-mirror/cluster_watcher.go

```diff
@@ -0,0 +1,610 @@
+	return gatewayEndpoints, gateway.Spec.Ports[0].Port, nil
+}
+
+// NewRemoteClusterServiceWatcher constructs a new cluster watcher
+func NewRemoteClusterServiceWatcher(localAPI *k8s.API, cfg *rest.Config, clusterName string, requeueLimit int) (*RemoteClusterServiceWatcher, error) {
+	remoteAPI, err := k8s.InitializeAPIForConfig(cfg, k8s.Svc)
+	if err != nil {
+		return nil, fmt.Errorf("cannot initialize remote api for cluster %s: %s", clusterName, err)
+	}
+	stopper := make(chan struct{})
+	return &RemoteClusterServiceWatcher{
+		clusterName:     clusterName,
+		remoteAPIClient: remoteAPI,
+		localAPIClient:  localAPI,
+		stopper:         stopper,
+		log: logging.WithFields(logging.Fields{
+			"cluster":    clusterName,
+			"apiAddress": cfg.Host,
+		}),
+		eventsQueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		requeueLimit: requeueLimit,
+	}, nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) mirroredResourceName(remoteName string) string {
+	return fmt.Sprintf("%s-%s", remoteName, rcsw.clusterName)
+}
+
+func (rcsw *RemoteClusterServiceWatcher) originalResourceName(mirroredName string) string {
+	return strings.TrimSuffix(mirroredName, fmt.Sprintf("-%s", rcsw.clusterName))
+}
+
+func (rcsw *RemoteClusterServiceWatcher) getMirroredServiceLabels(service *corev1.Service) map[string]string {
+	newLabels := map[string]string{
+		MirroredResourceLabel:      "true",
+		RemoteClusterNameLabel:     rcsw.clusterName,
+		RemoteResourceVersionLabel: service.ResourceVersion, // needed to detect real changes
+	}
+	for k, v := range service.Labels {
+		// rewrite the mirroring ones
+		if k == GatewayNameAnnotation {
+			k = RemoteGatewayNameLabel
+		}
+		if k == GatewayNsAnnotation {
+			k = RemoteGatewayNsLabel
+		}
+		newLabels[k] = v
+	}
+	return newLabels
+}
+
+func (rcsw *RemoteClusterServiceWatcher) mirrorNamespaceIfNecessary(namespace string) error {
+	// if the namespace is already present we do not need to change it.
+	// if we are creating it we want to put a label indicating this is a
+	// mirrored resource so we delete it when the times comes...
+	if _, err := rcsw.localAPIClient.NS().Lister().Get(namespace); err != nil {
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					MirroredResourceLabel:  "true",
+					RemoteClusterNameLabel: rcsw.clusterName,
+				},
+				Name: namespace,
+			},
+		}
+		_, err := rcsw.localAPIClient.Client.CoreV1().Namespaces().Create(ns)
+		if err != nil {
+			return fmt.Errorf("could not create mirror namespace: %s", err)
+		}
+	}
+	return nil
+}
+
+// This method takes care of port remapping. What it does essentially is get the one gateway port
+// that we should send traffic to and create endpoint ports that bind to the mirrored service ports
+// (same name, etc) but send traffic to the gateway port. This way we do not need to do any remapping
+// on the service side of things. It all happens in the endpoints.
```

Comment:
I might be a bit unfamiliar with the k8s concepts here, so do you mind putting that in a concrete example. So currently if we have a remote service that has ports 8080 and 8081 defined and we have a gateway that has port 443 on it, we would have the following configuration on the local side: 

```yaml
apiVersion: v1
kind: Service
metadata:
  name: foobar-remote
  namespace: default
spec:
  ports:
  - name: port-name-1
    port: 8080
    protocol: TCP
  - name: port-name-2
    port: 8081
    protocol: TCP

apiVersion: v1
kind: Endpoints
metadata:
  name: foobar-remote
  namespace: default
subsets:
- addresses:
  - ip: <some-gateway-ip>
  ports:
  - name: port-name-1
    port: 443
    protocol: TCP
  - name: port-name-2
    port: 443
    protocol: TCP
```
So in a sense our ports configuration on the service is identical as the one in the remote service. I am not sure what kind of change you are suggesting. I was under the impression that the port remapping on the service side pertains only to containers that this service is targeting (i.e. you hit port 8080 on the service but you really want that to target port 7777 on the container). Might be getting all this wrong, but was under the impression that this is what we need here.

Reply:
In your example, I think the local configuration should be:

```
apiVersion: v1
kind: Service
metadata:
  name: foobar-remote
  namespace: default
spec:
  ports:
  - name: port-name-1
    port: 8080
    targetPort: 443
    protocol: TCP
  - name: port-name-2
    port: 8081
    targetPort: 443
    protocol: TCP

apiVersion: v1
kind: Endpoints
metadata:
  name: foobar-remote
  namespace: default
subsets:
- addresses:
  - ip: <some-gateway-ip>
```

Notice that we apply the port mapping in the service.  The endpoints themselves don't need to define any named ports, because we don't use any named ports.

Reply:
@adleong Will try that out. Was not aware that this actually works. Will give it a shot. Thanks !

Reply:
@adleong I do not think that works the way I expect it. So for example I have the following resources in a local kind cluster: 


```
apiVersion: v1
kind: Service
metadata:
  name: foobar-remote
  namespace: default
spec:
  ports:
  - name: http
    port: 8080
    protocol: TCP
    targetPort: 80
  - name: https
    port: 8443
    protocol: TCP
    targetPort: 443
---
apiVersion: v1
kind: Endpoints
metadata:
  name: foobar-remote
  namespace: default
subsets:
- addresses:
  - ip: 194.153.145.104 
```
`194.153.145.104 ` is a popular Bulgarian website that can be resolved via `abv.bg`. So from inside a container, this is what is happening: 

I can do  `curl https://194.153.145.104:443`
I can do  `curl https://abv.bg:443`
I **cannot** do `curl https://foobar-remote.default.svc.cluster.local:8443`

I think the port mapping needs to happen on the Endpoints after all, but I might be missing something here... Also the documentation mentions that the "targetPort" om the ServicePorts pertains only to targeting ports on pods. @grampelberg do you see something we are missing here?

Reply:
That behavior is... surprising to me.  The docs are a bit ambiguous about this:

> targetPort: Number or name of the port to access on the pods targeted by the service.

Except that a service doesn't always target pods 😞.

I'm looking at this example: https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors

Perhaps we need both?  In other words, we need to perform the port mapping in the service, but the endpoint also needs to define the target port so that the port mapping in the service has something concrete to target?  :shrug:

Reply:
@adleong @grampelberg  So I dug a bit more into the k8s source and came to the conclusion that we are fine using the endpoints mapping. I think this ties together better with the whole update/create mechanism as in this cse we only need to touch the endpoints object in case we have some update on the gateway. Futhermore, I think this piece of code validates my assumption https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go#L42. 

If I am not mistaken, this code does the rempaping and looks for named ports (if more than one) and simply maps the target port to the port in the endpoints subsets. So far my tests validate that this is what happens.

Reply:
After doing some more digging and testing, it seems that you are correct.  I had the wrong mental model for how this works.  We actually fixed Linkerd's handling of this, but I wasn't aware: https://github.com/linkerd/linkerd2/pull/1973

The logic, when resolving a request seems to be:
* If making a request to port X, search the service object for a port spec with port=X
* Then search the endpoints object for a port with the same name as the port and use that port number
* (the targetPort field of the service is completely ignored)

It seems that the targetPort is only used for the purposes of creating the endpoints object and is not used during resolution.  Since we're creating endpoints objects manually, we don't have to worry about it at all.  But I think we should still set it to the correct value for consistency's sake and so that our endpoints object and service object match up.

TLDR: you were totally right about all of this.

Reply:
The problem is that then we need to take care of updating more than one things in certain situations and I would like to avoid it. A clear example of that is.. Our gateway updates, and its port is updated (for some bizzare reason). Now we need to go update both the service and endpoints definitions. Otherwise, we just update the endpoint. So if its only for consistency's sake, I am not sure its worth it. But yes, agree this logic should be documented a lot better on the k8s side. Took me a lot of trial and error to arrive at the conclusion until I looked at the codebase.